pandas/io/formats/xml.py

```diff
@@ -0,0 +1,674 @@
+            if isinstance(col, tuple):
+                flat_col = (
+                    "".join(str(c) for c in col).strip()
+                    if "" in col
+                    else "_".join(str(c) for c in col).strip()
+                )
+
+            attr_name = f"{self.prefix_uri}{flat_col}"
+            try:
+                val = (
+                    None
+                    if self.d[col] is None or self.d[col] != self.d[col]
+                    else str(self.d[col])
+                )
+                if val is not None:
+                    self.elem_row.attrib[attr_name] = val
+            except KeyError:
+                raise KeyError(f"no valid column, {col}")
+
+    def build_elems(self) -> None:
+        from lxml.etree import SubElement
+
+        if not self.elem_cols:
+            return
+
+        for col in self.elem_cols:
+            flat_col = col
+            if isinstance(col, tuple):
+                flat_col = (
+                    "".join(str(c) for c in col).strip()
+                    if "" in col
+                    else "_".join(str(c) for c in col).strip()
+                )
+
+            elem_name = f"{self.prefix_uri}{flat_col}"
+            try:
+                val = (
+                    None
+                    if self.d[col] in [None, ""] or self.d[col] != self.d[col]
+                    else str(self.d[col])
+                )
+                SubElement(self.elem_row, elem_name).text = val
+            except KeyError:
+                raise KeyError(f"no valid column, {col}")
+
+    def parse_doc(self):
+        """
+        Build tree from stylesheet.
+
+        This method will parse stylesheet object into tree for parsing
+        conditionally by its specific object type.
+        """
+
+        from lxml.etree import (
+            XMLParser,
+            fromstring,
+            parse,
+        )
+
+        style_doc = self.stylesheet
+
+        handle_data = _get_data_from_filepath(
+            filepath_or_buffer=style_doc,
+            encoding=self.encoding,
+            compression=self.compression,
+            storage_options=self.storage_options,
+        )
+
+        with _preprocess_data(handle_data) as xml_data:
+            curr_parser = XMLParser(encoding=self.encoding)
+
+            if isinstance(xml_data, io.StringIO):
+                r = fromstring(
+                    xml_data.getvalue().encode(self.encoding), parser=curr_parser
+                )
+            else:
+                r = parse(xml_data, parser=curr_parser)
+
+        return r
+
+    def transform_doc(self) -> bytes:
+        """
+        Transform original tree using stylesheet.
+
+        This method will transform built tree with XSLT script.
+        """
+
+        from lxml.etree import XSLT
+
+        xsl_doc = self.parse_doc()
+
+        transformer = XSLT(xsl_doc)
+        new_doc = transformer(self.root)
+
+        return bytes(new_doc)
+
+
+def _get_data_from_filepath(
```

Comment:
This looks to be one of the only functions not typed; always nice to have (can be done in a follow up)

Reply:
Also is this a copy of the function in the other `xml.py` module? Would be nice to de-duplicate

Reply:
Re `parse_doc` not typed in Lxml classes, originally I had it typed but lxml unlike etree has its `_Element` and `_ElementTree` objects as private variables (with leading underscores) which fails mypy on import. Can ignore.

Yes, methods do repeat. Can pull out of class as module level method in pandas.io.import (i.e., `read_xml`) to be imported for in pandas.io.formats.xml (i.e., `to_xml`).

Reply:
Digging deeper, the private variables were from modified class objects. lxml does use same named types as etree. Adjusted accordingly.